src/braket/default_simulator/cirq/cirq_program_context.py

```diff
@@ -0,0 +1,122 @@
+    GeneralizedAmplitudeDampingChannel,
+    PhaseDampingChannel,
+)
+
+from braket.default_simulator.openqasm.parser.braket_pragmas import (
+    AbstractBraketPragmaNodeVisitor,
+    parse_braket_pragma,
+)
+from braket.default_simulator.openqasm.parser.generated.BraketPragmasParser import (
+    BraketPragmasParser,
+)
+from braket.default_simulator.openqasm.program_context import AbstractProgramContext
+
+# cirq.XX, cirq.YY, and cirq.ZZ gates are not the same as Braket gates
+CIRQ_GATES = {
+    "i": cirq.I,
+    "h": cirq.H,
+    "x": cirq.X,
+    "y": cirq.Y,
+    "z": cirq.Z,
+    "cnot": cirq.CNOT,
+    "cz": cirq.CZ,
+    "s": cirq.S,
+    "t": cirq.T,
+    "cphaseshift": cirq.cphase,
+    "rx": cirq.rx,
+    "ry": cirq.ry,
+    "rz": cirq.rz,
+    "swap": cirq.SWAP,
+    "iswap": cirq.ISWAP,
+    "ccnot": cirq.CCNOT,
+    "cswap": cirq.CSWAP,
+    "ccz": cirq.CCZ,
+    "ccx": cirq.CCX,
+    "measure": cirq.MeasurementGate,
+}
+
+
+class CirqProgramContext(AbstractProgramContext):
+    def __init__(self):
+        super().__init__(Circuit())
+
+    def is_builtin_gate(self, name: str) -> bool:
+        user_defined_gate = self.is_user_defined_gate(name)
+        return name in CIRQ_GATES and not user_defined_gate
+
+    def add_phase_instruction(self, target: Tuple[int], phase_value: int):
+        raise NotImplementedError
+
+    def _gate_accepts_parameters(self, gate_class):
+        return hasattr(gate_class, "parameters") and len(gate_class.parameters) > 0
+
+    def add_gate_instruction(
+        self, gate_name: str, target: Tuple[int], params, ctrl_modifiers: List[int], power: int
+    ):
+        qubits = [cirq.LineQubit(int(qubit)) for qubit in target]
+        target_qubits = qubits[len(ctrl_modifiers) :]
+        control_qubits = qubits[: len(ctrl_modifiers)]
+
+        if params:
+            gate = CIRQ_GATES[gate_name](*params).on(*target_qubits)
+        else:
+            gate = CIRQ_GATES[gate_name].on(*target_qubits)
+        ctrl_modifiers = [bit ^ 1 for bit in ctrl_modifiers]
+        gate = gate.controlled_by(*control_qubits, control_values=ctrl_modifiers)
+        gate = gate**power
+
+        self.circuit.append(gate)
+
+    def add_custom_unitary(
+        self,
+        unitary: np.ndarray,
+        target: Tuple[int],
+    ) -> None:
+        """Add a custom Unitary instruction to the circuit"""
+        raise NotImplementedError
+
+    def add_noise_instruction(self, noise):
+        """Add a noise instruction the circuit"""
+        self.circuit.append(noise)
+
+    def add_result(self, result: Results) -> None:
+        """Add a result type to the circuit"""
+        raise NotImplementedError
+
+    def parse_pragma(self, pragma_body: str):
+        """Parse pragma"""
+        return parse_braket_pragma(pragma_body, CirqBraketPragmaNodeVisitor(self.qubit_mapping))
+
+
+class CirqBraketPragmaNodeVisitor(AbstractBraketPragmaNodeVisitor):
+    def visitNoise(self, ctx: BraketPragmasParser.NoiseContext):
```

Comment:
I'm a bit confused why we need to make changes to the visitor now. The existing visitor should be capturing all noise instructions into the circuit. For gates, you just mapped gates from the braket sim circuit to cirq; can we do the same here?

Reply:
Since noise gates are a part of Visitor class. I had to make a decision to either abstract it or move the code to the Program Context.

Due to backward compatibility, I had to do it this way.

The user has to define AbstractProgramContext and AbstractBraketPragmaNodeVisitor to define the gates

Since noise gates are present in Visitor and all the other basic gates in ProgramContext